### PR TITLE
perf(qwen3.8): 3-stage cp.async pipeline for the FP8 GDN prefill projections (1.03x prefill@128)

### PR DESCRIPTION
**Pull request by FranDev132**
## Summary

The GDN in-projections (`in_proj_qkv`, `in_proj_z`, `out_proj`, 48 layers) are the largest single
group left in prefill@128 — 6.40 ms of a 26.16 ms pass — and they move 5.54 GB of e4m3 weight at
only ~48% of this part's bandwidth. The cause is pipeline depth, not tiling: `pf_gemm_fp8_kernel`
runs a **2-stage** cp.async double buffer, so exactly one prefetch is ever in flight, and its
accumulators (`acc[2][8][4]` + `h[2][8][2]`) hold `__launch_bounds__(256, 2)` to 8 warps × 2 blocks
with no other warp resident to cover the miss. At 1.1–1.4 waves these shapes have no neighbouring
block to hide latency either, so the K loop pays a full global latency on nearly every trip.

This takes the pipeline to 3 stages. Depth costs shared memory, not registers, so occupancy is
unchanged: 3 × 16 KB = 48 KB, still two blocks per SM. Shared memory moves to dynamic so the depth
is a template parameter rather than a second kernel.

Only the **issue point** of the staging moves — the mma order and the `FP8_FLUSH` cadence are
untouched, so the result is bit-identical (verified below, not asserted).

Depth 4 measures *worse than the shipped 2* (4802 vs 4839 pp): 64 KB puts one block on an SM and
halves occupancy. 3 is the knee, and it is where the isolated kernel saturates too.

`SPARKINFER_PREFILL_FP8_STAGES=2` restores the shipped pipeline, and `=4` is kept instantiated, so
all three depths A/B from one binary.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end, `qwen3_gguf_bench <dir> 128 sweep`, `SWEEP_CTXS=128`, `REPS=5`):

| | decode tok/s |
|---|--:|
| before (main) | 84.35 |
| after (this PR) | 84.33 |

**Prefill pp tok/s** (same sweep, same model load, ctx=128 — the scored context):

| | prefill pp tok/s |
|---|--:|
| before prefill (main) | 4839.19 |
| after prefill (this PR) | 4971.01 |

**+2.72% prefill@128**, decode flat, VRAM unchanged at 30.8 GB. Same-box RTX 5090 vs `origin/main`
@ `e9b4a2c`, `unsloth/Qwen3.8-27B-NVFP4`, one binary with the depth selected by env, alternated,
5 reps inside each process, median of three rounds:

```text
  r1 stages=2 SWEEP_JSON {"128":{"decode_tps":84.4489,"prefill_pp":4839.3702}}
  r1 stages=3 SWEEP_JSON {"128":{"decode_tps":84.4205,"prefill_pp":4992.4883}}
  r1 stages=4 SWEEP_JSON {"128":{"decode_tps":84.3786,"prefill_pp":4830.7405}}
  r2 stages=2 SWEEP_JSON {"128":{"decode_tps":84.3455,"prefill_pp":4839.1854}}
  r2 stages=3 SWEEP_JSON {"128":{"decode_tps":84.3289,"prefill_pp":4971.0056}}
  r2 stages=4 SWEEP_JSON {"128":{"decode_tps":84.3002,"prefill_pp":4802.2496}}
  r3 stages=2 SWEEP_JSON {"128":{"decode_tps":84.2967,"prefill_pp":4821.7256}}
  r3 stages=3 SWEEP_JSON {"128":{"decode_tps":84.2757,"prefill_pp":4957.7090}}
  r3 stages=4 SWEEP_JSON {"128":{"decode_tps":84.2480,"prefill_pp":4791.4295}}
```

```text
# main (2-stage)
[compressed-tensors] loaded 64 layers, native NVFP4 prefill FFN 56/64 (decode stays Q4_K)
VRAM used    : 30.8 GB
decode tg    : 84.44 tok/s  (n=128, ctx=128, bs=1)
prefill pp   : 4834.37 tok/s  (ctx=128)
SWEEP_JSON {"128":{"decode_tps":84.4412,"prefill_pp":4834.3732}}

# this PR (3-stage)
[compressed-tensors] loaded 64 layers, native NVFP4 prefill FFN 56/64 (decode stays Q4_K)
VRAM used    : 30.8 GB
decode tg    : 84.41 tok/s  (n=128, ctx=128, bs=1)
prefill pp   : 4969.31 tok/s  (ctx=128)
SWEEP_JSON {"128":{"decode_tps":84.4072,"prefill_pp":4969.3056}}
```

Isolated kernel, for the mechanism only (not the before/after claim above) — M=128, non-split-K,
same three shapes, 300 iterations:

```text
qkv [10240,5120]   2-stage 74.19 us -> 3-stage 62.49 us  (+15.8%)
z   [ 6144,5120]   2-stage 73.81 us -> 3-stage 61.92 us  (+16.1%)
out [ 5120,6144]   2-stage 86.48 us -> 3-stage 72.72 us  (+15.9%)
```

## Correctness

Bit-identical, not merely within tolerance. `qwen3_gguf_prefill_check <dir> 128 64` with
`SPARKINFER_KV_INT8=0` (batched prefill vs the token loop over the same prompt, teacher-forced,
per-position logits) returns the same value at every depth from one binary:

| depth | TOP1 | KL |
|---|--:|--:|
| 2 (main) | 0/64 | 7.83639 |
| 3 (this PR) | 0/64 | 7.83639 |
| 4 | 0/64 | 7.83639 |

`qwen3_gguf_score` dumps compare byte-identical between depth 2 and 3.

> The absolute TOP1/KL above are main's, unchanged by this PR — the ctx=128 batched-prefill path is
> already numerically divergent from the token loop on this model before this change. This PR does
> not move it in either direction; the point of the table is that all three depths agree exactly.

`qwen3_gguf_prefill_check` grows compressed-tensors directory support so this check can run on this
model at all — it opened GGUF only.

## Qwen3.6 no-regression

`launch_prefill_gemm_fp8` is shared with Qwen3.6's MoE prefill, so the same binary was swept there
(`SWEEP_CTXS=0,512,4096,16384,32768`, `REPS=2`, ratio = 3-stage / 2-stage):

| ctx | decode | prefill |
|---|--:|--:|
| 0 | 1.000 | — |
| 512 | 1.000 | 1.015 |
| 4096 | 1.001 | 1.002 |
| 16384 | 0.999 | 0.999 |
| 32768 | 0.998 | 0.999 |

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## sparkinfer qwen38 auto-eval — `eval-qwen38:none`

| metric | value |
|---|---|
| **label** | `eval-qwen38:none` |
| scored at | prefill@16k (the only scoring dimension); decode@128 + prefill@128 are no-regression floors |
| tier came from | `prefill@16k` |
| PR decode tok/s | 84.07 |
| main decode tok/s | 84.09 |
| decode speedup vs main | **1.00×** (-0.0%) |
| PR prefill@128 pp | 5206.01 |
| main prefill@128 pp | 5159.02 |
| prefill@128 vs main | +0.9% |
| PR prefill@16k pp | 79.82 |
| main prefill@16k pp | 79.84 |
| prefill@16k vs main | -0.0% |
| accuracy gate | ✅ top1=1.000 (bar >=0.99) · KL=0.0000 (bar <=0.01) |
| qwen3.6 guard | ✅ no regression (decode+prefill, ctx 0/512/4k/16k/32k) |
| PPL PR / main | 3.2812 / 3.2812 |
| Polaris receipt | `61db5c29af28e4c2…` — TDX (Intel hardware attestation) |
| commit | `5c8603774` |

within significance gate — not a verified prefill@16k improvement

<sub>Scored on the pinned eval box vs same-box `origin/main` — 128-token AR decode (ctx=0) AND 128-ctx prefill throughput, from one model load; either dimension regressing is a hard REJECT, but otherwise the reported label is the **better** of the two tiers — a PR that improves just one, with the other flat, still earns credit for that (no DFlash, no long-context beyond 128) — Qwen3.8-27B's narrow, deliberately strict eval scope. This is informational, not a judgment on your PR: a `none` label just means no measurable Qwen3.8-27B speedup was verified on either metric, which is expected and fine if that isn't what your change is about. Correctness gated against a live llama.cpp reference on the same GGUF. Also gated on a Qwen3.6 no-regression guard (decode+prefill, ctx 0/512/4k/16k/32k, same box vs main) — Qwen3.8-27B PRs can touch code shared with other models. Automated — **not merged**; merge manually after review.</sub>

**Comment by skyrocket2026** [comment]
## Closed: sparkinfer qwen38 auto-eval — `eval-qwen38:none`

This PR's Qwen3.8-27B prefill@16k speed measured **-0.0%** vs main, with no verified prefill@16k improvement (the only scored dimension) — closing automatically. This bot evaluates every eligible PR in the repo against Qwen3.8-27B's decode AND prefill@128 speed specifically, regardless of what the PR is actually about — a close here isn't a judgment on the PR's purpose, just that it didn't move these particular metrics. Reopen (or open a fresh PR) if you have a fix or a different approach.
